src/tests/ab_server/src/cip.c

```diff
@@ -562,7 +571,7 @@ slice_s handle_write_request(slice_s input, slice_s output, plc_s *plc)
     info("tag_data_length = %d", tag_data_length);
 
     /* get the write amount requested. */
-    total_request_size = slice_len(input) - offset;
+    total_request_size = slice_len(input) - offset - 1;
```

Comment:
I do not think this is correct.   Here are some tests I did:

1. A  write test with a tag named TstBigArray[2000] of type SINT (same size as BOOL).
2. A write test with a tag named TestBigArray[2000] of type DINT (note the different tag name).
3. A write test with a tag named TestBigArray[2000] of type SINT but with the access starting at element 9, so the tag name in the request is `TestBigArray[9]`.
4. A write test with a tag named TestBigArray[2000] of type DINT but with the access starting at element 9, so the tag name in the request is `TestBigArray[9]`.
5. A write test with a tag named TestBigArray[2000] of type DINT but with the access starting at element 9, so the tag name in the request is `TestBigArray[9]`.  This time I access 1000 elements.  This should cause fragmentation.

Test 1:
```
INFO cip_dispatch_request:97 Got packet:
000 4d 07 91 0b 54 73 74 42 69 67
010 41 72 72 61 79 00 c2 00 04 00
020 01 02 03 04
INFO process_tag_segment:635 Found tag TstBigArray
INFO process_tag_segment:647 Numeric segment(s):
INFO handle_write_request:559 byte_offset = 0
INFO handle_write_request:564 tag_data_length = 2000
INFO handle_write_request:569 total_request_size = 4
INFO handle_write_request:578 byte_offset = 0
INFO handle_write_request:579 offset = 20
INFO handle_write_request:580 total_request_size = 4
```
The `total_request_size` is correct here.

Test 2:
```
INFO cip_dispatch_request:97 Got packet:
000 4d 07 91 0c 54 65 73 74 42 69
010 67 41 72 72 61 79 c4 00 04 00
020 05 00 00 00 06 00 00 00 07 00
030 00 00 08 00 00 00
INFO process_tag_segment:635 Found tag TestBigArray
INFO process_tag_segment:647 Numeric segment(s):
INFO handle_write_request:559 byte_offset = 0
INFO handle_write_request:564 tag_data_length = 8000
INFO handle_write_request:569 total_request_size = 16
INFO handle_write_request:578 byte_offset = 0
INFO handle_write_request:579 offset = 20
INFO handle_write_request:580 total_request_size = 16
```

Again the request payload size is correct.  4 DINTs is 16 bytes.

Test 3:
```
INFO cip_dispatch_request:97 Got packet:
000 4d 08 91 0c 54 65 73 74 42 69
010 67 41 72 72 61 79 28 09 c2 00
020 04 00 01 02 03 04
INFO process_tag_segment:635 Found tag TestBigArray
INFO process_tag_segment:647 Numeric segment(s):
000 28 09
INFO handle_write_request:559 byte_offset = 0
INFO handle_write_request:564 tag_data_length = 2000
INFO handle_write_request:569 total_request_size = 4
INFO handle_write_request:578 byte_offset = 0
INFO handle_write_request:579 offset = 22
INFO handle_write_request:580 total_request_size = 4
```

The counts and sizes still look good.

Test 4:
```
INFO cip_dispatch_request:97 Got packet:
000 4d 08 91 0c 54 65 73 74 42 69
010 67 41 72 72 61 79 28 09 c4 00
020 04 00 0a 00 00 00 0b 00 00 00
030 0c 00 00 00 0d 00 00 00
INFO process_tag_segment:635 Found tag TestBigArray
INFO process_tag_segment:647 Numeric segment(s):
000 28 09
INFO handle_write_request:559 byte_offset = 0
INFO handle_write_request:564 tag_data_length = 8000
INFO handle_write_request:569 total_request_size = 16
INFO handle_write_request:578 byte_offset = 0
INFO handle_write_request:579 offset = 22
INFO handle_write_request:580 total_request_size = 16
```

Still good.

So I do not think that this change to the total request size is correct.  

Test 5:
```
INFO cip_dispatch_request:97 Got packet:
000 53 08 91 0c 54 65 73 74 42 69
010 67 41 72 72 61 79 28 09 c4 00
020 e8 03 00 00 00 00 14 00 00 00
030 15 00 00 00 16 00 00 00 17 00
...
3970 00 00 00 00 00 00 00 00 00 00
3980 00 00 00 00 00 00 00 00 00 00
3990 00 00 00 00
INFO process_tag_segment:635 Found tag TestBigArray
INFO process_tag_segment:647 Numeric segment(s):
000 28 09
INFO handle_write_request:559 byte_offset = 0
INFO handle_write_request:564 tag_data_length = 8000
INFO handle_write_request:569 total_request_size = 3968
INFO handle_write_request:578 byte_offset = 0
INFO handle_write_request:579 offset = 26
INFO handle_write_request:580 total_request_size = 3968
```
This looks correct. 3968 bytes is exactly 992 4-byte elements.    So we should see a second write request with 8 elements.


Here's the second write containing the remaining 8 elements.
```
INFO cip_dispatch_request:97 Got packet:
000 53 08 91 0c 54 65 73 74 42 69
010 67 41 72 72 61 79 28 09 c4 00
020 e8 03 80 0f 00 00 00 00 00 00
030 00 00 00 00 00 00 00 00 00 00
040 00 00 00 00 00 00 00 00 00 00
050 00 00 00 00 00 00 00 00
INFO process_tag_segment:635 Found tag TestBigArray
INFO process_tag_segment:647 Numeric segment(s):
000 28 09
INFO handle_write_request:559 byte_offset = 3968
INFO handle_write_request:564 tag_data_length = 8000
INFO handle_write_request:569 total_request_size = 32
INFO handle_write_request:578 byte_offset = 3968
INFO handle_write_request:579 offset = 26
INFO handle_write_request:580 total_request_size = 32
```
This all looks correct to me.

Where did you find an error that needed that `-1`?  I am not seeing any problems with my tests above.

Reply:
There are problems with types only using 1 byte such as SINT or BOOL.
When I create tags with size 1 as following:
`ab_server --debug --plc=ControlLogix --path=1,0 --tag=SintExample:SINT[1] --tag=IntExample:INT[1] --tag=DintExample:DINT[1] --tag=LintExample:LINT[1] --tag=RealExample:REAL[1] --tag=LrealExample:LREAL[1] --tag=StringExample:STRING[1] --tag=BoolExample:BOOL[1]`

I use .NET Wrapper to write SINT value and ErrrorOutOfBounds happened:
`INFO cip_dispatch_request:108 Case CIP_WRITE
INFO process_tag_segment:642 Found tag SintExample
INFO process_tag_segment:654 Numeric segment(s):
INFO handle_write_request:566 byte_offset = 0
INFO handle_write_request:571 tag_data_length = 1
INFO handle_write_request:576 total_request_size = 2
INFO handle_write_request:580 request tries to write too much data!
INFO socket_write:291 socket_write(): writing packet:
000 70 00 1c 00 b0 78 00 00 00 00
010 00 00 00 00 00 00 00 00 00 00
020 00 00 00 00 02 00 a1 00 04 00
030 4e 3a 00 00 b1 00 08 00 00 00
040 b1 00 18 00 0c 00 cd 00 ff 02
050 05 21
INFO eip_dispatch_request:74 eip_dispatch_request(): got packet:
000 6f 00 28 00 b0 78 00 00 00 00
010 00 00 18 4f 00 00 00 00 00 00
020 00 00 00 00 02 00 a1 00 01 00
030 02 00 00 00 00 00 b2 00 18 00
040 4e 02 20 06 24 01 0a 05 0a 0d
050 3d f3 45 43 50 21 03 00 01 00
060 20 02 24 01
`

This also happens when using BOOL tag:
`INFO cip_dispatch_request:108 Case CIP_WRITE
INFO process_tag_segment:642 Found tag BoolExample
INFO process_tag_segment:654 Numeric segment(s):
INFO handle_write_request:566 byte_offset = 0
INFO handle_write_request:571 tag_data_length = 1
INFO handle_write_request:576 total_request_size = 2
INFO handle_write_request:580 request tries to write too much data!
INFO socket_write:291 socket_write(): writing packet:
000 70 00 1c 00 f8 12 00 00 00 00
010 00 00 00 00 00 00 00 00 00 00
020 00 00 00 00 02 00 a1 00 04 00
030 4e 3a 00 00 b1 00 08 00 00 00
040 b1 00 18 00 0e 00 cd 00 ff 02
050 05 21
INFO eip_dispatch_request:74 eip_dispatch_request(): got packet:
000 6f 00 28 00 f8 12 00 00 00 00
010 00 00 1a 4f 00 00 00 00 00 00
020 00 00 00 00 02 00 a1 00 01 00
030 02 00 00 00 00 00 b2 00 18 00
040 4e 02 20 06 24 01 0a 05 0b 0d
050 3d f3 45 43 50 21 03 00 01 00
060 20 02 24 01
INFO handle_cpf_unconnected:79 handle_cpf_unconnected(): got packet:
000 02 00 a1 00 01 00 02 00 00 00
010 00 00 b2 00 18 00 4e 02 20 06
020 24 01 0a 05 0b 0d 3d f3 45 43
030 50 21 03 00 01 00 20 02 24 01
INFO cip_dispatch_request:97 Got packet:
000 4e 02 20 06 24 01 0a 05 0b 0d
010 3d f3 45 43 50 21 03 00 01 00
020 20 02 24 01
`

This is why I think the calculation of total_request_size is wrong.
However, `total_request_size = slice_len(input) - offset` is right for other types using more than 1 byte so I think we should check the calculation of total request size when write again.

Reply:
Hmm, I think I see the problem.  The write request is the wrong length coming from libplctag.  It looks like if the write length is an odd number of buyers then a padding byte is added. 

@ptsOSL did some changes for Omron PLCs and he found this was a problem with strings.  There is a config attribute to set this in the prerelease branch. 

Subtracting one does not actually fix the real problem.  This code has been working for years so the real PLCs must just truncate the write payload.  

I need to think about how to fix this.  

Can you test with odd numbers of array elements with a SINT or BOOL?  If the problem is what I think then all odd length payloads might hit this.

Reply:
When testing with Omron NJ, adding write requests was an issue for string requests because the null bytes changed the length of the string. For counted strings which include a 2 byte int which counts the number of characters, this caused an issue as this count differed from the length of the string. I added the str_pad_to_multiple_bytes parameter to allow you to change whether write requests are padded to 1 byte, 2 bytes or 4. Setting this to 1 stops this issue with the strings.

For SINT and USINT I did not have this problem, the PLC just ignores the padding, but I do think that the driver is technically incorrect to send this padding. When you read a SINT/USINT, the PLC does not pad the data. It sends just three bytes, with the first two bytes containing the datatype code and the third byte containing the data.

I can imagine that other PLCs may not be so forgiving with this extra padding and this may result in errors. If this is an issue, then setting str_pad_to_multiple_bytes=1 for writing SINT and USINT data should fix the issue, at least for single writes, perhaps not for writing arrays.

Reply:
I tested odd numbers of array elements with a SINT and BOOL type, this is the command I use to open the server.
`ab_server --debug --plc=ControlLogix --path=1,0 --tag=SintExample:SINT[<k>] --tag=BoolExample:BOOL[<k>]` with odd number `k`

I use .NET Wrapper same as before.

**For `k = 1`,**

```INFO cip_dispatch_request:97 Got packet:
000 4d 07 91 0b 42 6f 6f 6c 45 78
010 61 6d 70 6c 65 00 c1 00 01 00
020 ff 00
INFO cip_dispatch_request:108 Case CIP_WRITE
INFO process_tag_segment:642 Found tag BoolExample
INFO process_tag_segment:654 Numeric segment(s):
INFO handle_write_request:566 byte_offset = 0
INFO handle_write_request:571 tag_data_length = 1
INFO handle_write_request:576 total_request_size = 2
INFO handle_write_request:580 request tries to write too much data!```

```INFO cip_dispatch_request:97 Got packet:
000 4d 07 91 0b 53 69 6e 74 45 78
010 61 6d 70 6c 65 00 c2 00 01 00
020 01 00
INFO cip_dispatch_request:108 Case CIP_WRITE
INFO process_tag_segment:642 Found tag SintExample
INFO process_tag_segment:654 Numeric segment(s):
INFO handle_write_request:566 byte_offset = 0
INFO handle_write_request:571 tag_data_length = 1
INFO handle_write_request:576 total_request_size = 2
INFO handle_write_request:580 request tries to write too much data!```

**For `k = 3`,**

```INFO cip_dispatch_request:97 Got packet:
000 4d 07 91 0b 42 6f 6f 6c 45 78
010 61 6d 70 6c 65 00 c1 00 01 00
020 ff 00
INFO cip_dispatch_request:108 Case CIP_WRITE
INFO process_tag_segment:642 Found tag BoolExample
INFO process_tag_segment:654 Numeric segment(s):
INFO handle_write_request:566 byte_offset = 0
INFO handle_write_request:571 tag_data_length = 3
INFO handle_write_request:576 total_request_size = 2```

```INFO cip_dispatch_request:97 Got packet:
000 4d 07 91 0b 53 69 6e 74 45 78
010 61 6d 70 6c 65 00 c2 00 01 00
020 01 00
INFO cip_dispatch_request:108 Case CIP_WRITE
INFO process_tag_segment:642 Found tag SintExample
INFO process_tag_segment:654 Numeric segment(s):
INFO handle_write_request:566 byte_offset = 0
INFO handle_write_request:571 tag_data_length = 3
INFO handle_write_request:576 total_request_size = 2```

**For `k = 5`,**

```INFO cip_dispatch_request:97 Got packet:
000 4d 07 91 0b 42 6f 6f 6c 45 78
010 61 6d 70 6c 65 00 c1 00 01 00
020 ff 00
INFO cip_dispatch_request:108 Case CIP_WRITE
INFO process_tag_segment:642 Found tag BoolExample
INFO process_tag_segment:654 Numeric segment(s):
INFO handle_write_request:566 byte_offset = 0
INFO handle_write_request:571 tag_data_length = 5
INFO handle_write_request:576 total_request_size = 2```

```INFO cip_dispatch_request:97 Got packet:
000 4d 07 91 0b 53 69 6e 74 45 78
010 61 6d 70 6c 65 00 c2 00 01 00
020 01 00
INFO cip_dispatch_request:108 Case CIP_WRITE
INFO process_tag_segment:642 Found tag SintExample
INFO process_tag_segment:654 Numeric segment(s):
INFO handle_write_request:566 byte_offset = 0
INFO handle_write_request:571 tag_data_length = 5
INFO handle_write_request:576 total_request_size = 2```

I found that `total_request_size` was always `2`.
The input packet was the same for all cases.
Adding a padding for SINT and BOOL type could be temporary solution for this problem, but I don't think this solves the problem.